### ISSUE TITLE
ULID implementation is not monotonic

### ISSUE THREAD

**Issue by robzienert** [huxi/sulky#3]
According to the spec, when multiple ULIDs are generated within the same millisecond, the least significant bits should be incremented, however it looks like `random.nextLong()` is always used, which causes ULIDs to be out of order.

Would you accept a PR to update the implementation to fix this, or have any input on implementation? I was going to do something simple like using an `AtomicInt` that would be incremented / reset on a millisecond boundary.

**Comment by huxi** [comment]
Yes, I would accept a PR but the functionality would have to be optional and mustn't influence the current performance, i.e. it would have to be implemented in a separate method like `public Value nextMonotonicValue()`.

This is all surprisingly non-trivial.

First of all, `System.currentTimeMillis()` is not guaranteed to provide ms precision. See this [question](https://stackoverflow.com/questions/7859019/system-currenttimemillis-is-not-accurate-on-windows-xp), for example. This seems to be fixed now on macOS 10.12.6 and Java 1.8.0_172 (I just checked and it finally has ms precision) but there is still no guarantee in general.

Another problem is the behavior in case of overflow... the current spec says this:

> If, in the extremely unlikely event that, you manage to generate more than 2<sup>80</sup> ULIDs within the same millisecond, or cause the random component to overflow with less, the generation will fail.

```js
ulid() // 01BX5ZZKBKZZZZZZZZZZZZZZZY
ulid() // 01BX5ZZKBKZZZZZZZZZZZZZZZZ
ulid() // throw new Error()!
```

I'm not convinced that this is desired behavior.

First of all, one does not need to generate 2<sup>80</sup> ULIDs within the same millisecond for this error to occur. One just needs "bad luck" regarding the initial random number obtained at the last ms boundary. And if Java does not provide millisecond accuracy in `System.currentTimeMillis()` (which is still, by API definition, system dependent), the chances for spurious `RuntimeException` would rise.

The only way to at least make sure users of the code are aware of this problem would be a checked exception and changing the signature to something like `public Value nextMonotonicValue() throws MonotonicValueOverflowException`. And what should be the recommended way to handle such an exception? Sleeping for 1ms?

In my opinion, generating a ULID should never cause an exception (at least not before `+10889-08-02T05:31:50.655Z` ^^) so an implementation of monotonic ULID values, if desired, should rather wait for the next millisecond in case of overflow instead of throwing an exception. Simply because throwing an exception would put the burden of handling it on the user of the API and chances are high that this handling wouldn't be implemented in a reasonable way.

But that's just my opinion and the spec defines a different behavior... What do you think about it?

The [original spec I used (see this fork from 2016)](https://github.com/huxi/ulid) didn't include the requirement for monotonic values at all. History was unfortunately lost when the spec migrated from`alizain/ulid` to [ulid/spec](https://github.com/ulid/spec/).

To implement this correctly, you'd have to keep the last milliseconds (for boundary checks) and the last two random `long` values. Then you'd have to add to the `long` value with the least significant bits and handle proper overflow into the `long` value with the most significant bits while also detecting and handling the 80 bits overall overflow. This means that simply using `AtomicLong` wouldn't be sufficient and calculation would need to be synchronized in a different way.

**Comment by robzienert** [comment]
Sorry for ghosting on this issue! Thanks so much for your detailed reply. From our use case, waiting until the next millisecond isn't exciting, but also seems the best given the constraints - I agree this edge case shouldn't be exposed to users.

Our app currently has a workaround for this specific issue, so it's not a blocker for us, but I'd still like to revisit in the future. If you don't mind, I'd like to keep the issue open but you can assign it to me if that makes sense.

**Comment by huxi** [comment]
@robzienert Take a look at the commit fixing this issue and let me know if this helps your use case.

I decided to put the burden of keeping track of the previous ULID value on the caller. That way the scope can be different for the various use cases lurking around.

**Comment by robzienert** [comment]
This looks like it'll fit the bill for us, we're perfectly happy to track the previous ULID in our code. Thanks a lot for putting this work together.

**Comment by huxi** [comment]
@robzienert I released 8.2.0 including those changes. Cheers!
